### DISCUSSION TITLE
Lutron Aurora out of sync with Home Assistant

### DISCUSSION THREAD

**Discussion by poblabs** [Koenkk/zigbee2mqtt#12511] [Devices]
Hi all, I have a question and I'm not sure how to solve it. I am using the Lutron Aurora dimmer and it's working great. However when pressing the button, the payload is either 0 or 255. Which is fine if the dimmer is the only thing controlling the lights. 

The issue is:

1. If I press the button, and 255 is sent, the lights turn on. 
2. If then use home assistant to turn off the lights.
3. I press the button again (say the next day) to turn the lights on, the payload is 0 because it's now out of sync from the actual light status. Which results in me having to press the button again. 

I am using node-red for my automations and can try to script something there, but if the payload is either 0 or 255, i don't think a script will help? 

Is there a way to get the device to send a "toggle" message on button press?

Any tips on how to try and get these into sync regardless of who or what is controlling the light status?

**Comment by rintaun** [comment]
I'm having this same issue, and it's quite frustrating. There is a workaround to determine button presses, but it has its own issues.

**The workaround:**
1. Set `legacy = false` in the device settings
2. You now get `brightness_move_to_level` for the action
3. `msg.data.new_state.attributes.action_transition_time` will consistently be `0.02` if it's rotation, and `0.07` if it's a button press. (At least, this is true for mine, ymmv I guess -- but they should be different I assume)

The _problem_ with that workaround, and using the button press as a pure toggle, is when you are controlling multiple lights.

For example, if you have an area with 2 lights, and you turn one of those lights (Light 1) `on` (e.g. via HA) while leaving the other (Light 2) `off`, when you toggle they will both swap, and Light 1 will now be `off`, with Light 2 now `on`.

This isn't a problem when using the Hue Bridge directly, so it must either be:
1. doing something extra to figure out whether to turn on or off
2. syncing the current state of the lights back to the dimmer

It might be possible to do some scripting and determine the state of the target lights, then turn on/off accordingly, but my frustration is that I can't find a way to generalize it. I currently have a `Lutron Aurora` subflow that I use whenever I have the dimmers, which outputs `turn_on`, `turn_off`, `brighten`, and `darken` so the caller use the relevant service for the correct area. But that has llimited utility if I have to do a bunch of logic outside the subflow (i.e. so i know which area to interact with).

**Reply by rintaun** [reply]
Using the HA `API` node I was able to get it working, kind of, with areas.

```
[{"id":"429ede05afcd50f8","type":"subflow","name":"Lutron Aurora","info":"","category":"","in":[{"x":200,"y":260,"wires":[{"id":"291959c26457ac85"}]}],"out":[{"x":2200,"y":220,"wires":[{"id":"e5a71e6455ca77c2","port":0}]},{"x":2200,"y":260,"wires":[{"id":"92300d0b246fee3e","port":0}]},{"x":2190,"y":300,"wires":[{"id":"73b8888a681e6ac0","port":0}]}],"env":[{"name":"area","type":"str","value":"","ui":{"label":{"en-US":"Area ID"},"type":"input","opts":{"types":["str","env"]}}}],"meta":{},"color":"#DDAA99","outputLabels":["Turn On","Turn Off","Dim"]},{"id":"291959c26457ac85","type":"switch","z":"429ede05afcd50f8","name":"Press or Rotate","property":"data.new_state.attributes.action_transition_time","propertyType":"msg","rules":[{"t":"eq","v":"0.07","vt":"str"},{"t":"eq","v":"0.02","vt":"str"}],"checkall":"false","repair":false,"outputs":2,"x":320,"y":260,"wires":[["609edde67e434a7c"],["73b8888a681e6ac0"]],"outputLabels":["Press","Rotate"]},{"id":"609edde67e434a7c","type":"ha-api","z":"429ede05afcd50f8","name":"Get entities for area","server":"31974050.8694b","version":1,"debugenabled":false,"protocol":"websocket","method":"get","path":"","data":"{\t   \"type\": \"search/related\",\t   \"item_type\": \"area\",\t   \"item_id\": $env('area')\t}","dataType":"jsonata","responseType":"json","outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"results"}],"x":530,"y":240,"wires":[["d9f5ee3050a4716d"]]},{"id":"c209141811abc60e","type":"ha-api","z":"429ede05afcd50f8","name":"Get light status","server":"31974050.8694b","version":1,"debugenabled":false,"protocol":"http","method":"get","path":"states/{{payload}}","data":"{}","dataType":"json","responseType":"json","outputProperties":[{"property":"payload","propertyType":"msg","value":"","valueType":"results"}],"x":1160,"y":220,"wires":[["a383d55a216e0a36"]]},{"id":"d9f5ee3050a4716d","type":"change","z":"429ede05afcd50f8","name":"","rules":[{"t":"set","p":"payload","pt":"msg","to":"payload.entity","tot":"jsonata"}],"action":"","property":"","from":"","to":"","reg":false,"x":720,"y":240,"wires":[["c69fca0ae9ae889f"]]},{"id":"c69fca0ae9ae889f","type":"split","z":"429ede05afcd50f8","name":"","splt":"\\n","spltType":"str","arraySplt":1,"arraySpltType":"len","stream":false,"addname":"","x":870,"y":240,"wires":[["4d6b9f174f1b5255"]]},{"id":"4d6b9f174f1b5255","type":"switch","z":"429ede05afcd50f8","name":"","property":"payload","propertyType":"msg","rules":[{"t":"regex","v":"^light\\.","vt":"str","case":true},{"t":"else"}],"checkall":"false","repair":false,"outputs":2,"x":990,"y":240,"wires":[["c209141811abc60e"],["7b996c96ab5f4264"]]},{"id":"a383d55a216e0a36","type":"join","z":"429ede05afcd50f8","name":"","mode":"auto","build":"array","property":"payload","propertyType":"msg","key":"topic","joiner":"\\n","joinerType":"str","accumulate":false,"timeout":"","count":"","reduceRight":false,"reduceExp":"","reduceInit":"","reduceInitType":"num","reduceFixup":"","x":1330,"y":240,"wires":[["69211d7702d986fe"]]},{"id":"7b996c96ab5f4264","type":"change","z":"429ede05afcd50f8","name":"","rules":[{"t":"set","p":"payload","pt":"msg","to":"undefined","tot":"jsonata"}],"action":"","property":"","from":"","to":"","reg":false,"x":1160,"y":260,"wires":[["a383d55a216e0a36"]]},{"id":"69211d7702d986fe","type":"function","z":"429ede05afcd50f8","name":"Area state","func":"const anyOff = msg.payload.reduce(\n    (res, cur) => {\n        if (res || !cur) { return res; }\n        return cur.state === 'off';\n    },\n    false\n);\n\nmsg.payload = anyOff ? 'turn_on' : 'turn_off';\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":1460,"y":240,"wires":[["ba2efeeb8d54fd80"]],"outputLabels":["Action"],"info":"If any lights are off, state will be \"off\".\nIf all lights are on, state will be \"on\"."},{"id":"73b8888a681e6ac0","type":"api-call-service","z":"429ede05afcd50f8","name":"Brighten or darken the lights","server":"31974050.8694b","version":5,"debugenabled":false,"domain":"light","service":"turn_on","areaId":[],"deviceId":[],"entityId":[],"data":"{\"area_id\": $env('area'), \"brightness\": msg.data.new_state.attributes.action_level, \"transition\": 0.5}","dataType":"jsonata","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","x":560,"y":280,"wires":[[]]},{"id":"92300d0b246fee3e","type":"api-call-service","z":"429ede05afcd50f8","name":"Turn off the lights","server":"31974050.8694b","version":5,"debugenabled":false,"domain":"light","service":"turn_off","areaId":[],"deviceId":[],"entityId":[],"data":"{\"area_id\": $env('area'), \"transition\": 0.2}","dataType":"jsonata","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","x":1770,"y":260,"wires":[[]]},{"id":"ba2efeeb8d54fd80","type":"switch","z":"429ede05afcd50f8","name":"Action","property":"payload","propertyType":"msg","rules":[{"t":"eq","v":"turn_on","vt":"str"},{"t":"eq","v":"turn_off","vt":"str"}],"checkall":"false","repair":false,"outputs":2,"x":1590,"y":240,"wires":[["e5a71e6455ca77c2"],["92300d0b246fee3e"]]},{"id":"e5a71e6455ca77c2","type":"api-call-service","z":"429ede05afcd50f8","name":"Turn on the lights","server":"31974050.8694b","version":5,"debugenabled":false,"domain":"light","service":"turn_on","areaId":[],"deviceId":[],"entityId":[],"data":"{\"area_id\": $env('area'), \"brightness_pct\": 100, \"transition\": 0.2}","dataType":"jsonata","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","x":1770,"y":220,"wires":[[]]},{"id":"31974050.8694b","type":"server","name":"Home Assistant","version":2,"addon":true,"rejectUnauthorizedCerts":true,"ha_boolean":"y|yes|true|on|home|open","connectionDelay":true,"cacheJson":true,"heartbeat":false,"heartbeatInterval":30}]
```

It takes the Lutron Aurora `action` message as input and has an environment variable for the area to use. Unfortunately, it's definitely not up to the `Call service` standards -- it only supports a single area, can't be tied to arbitrary entities, etc.

**Reply by poblabs** [reply]
This all seems promising but I don't have `sg.data.new_state.attributes.action_transition_time` when setting legacy to false.

**Reply by rintaun** [reply]
Hmm, that's unfortunate... I'll have to check my others to see the shape of the messages they generate.

Just to clarify, I'm using the  `events: state` node for the Aurora `_action` entity.

**Reply by rintaun** [reply]
Yeah, just checked and all 7 of my Auroras (purchased at different times from different places) have that shape with `legacy` set to `false`, when using the `_action` sensor. They get the state `brightness_move_to_level` with a bunch of different attributes compared to when `legacy` is `true`.

**Reply by poblabs** [reply]
bummer. any other ideas for a work around? 

@Koenkk any thoughts on this issue?
